_posts/2018-07-20-totw-153.md

```diff
@@ -0,0 +1,211 @@
+  ...
+}
+
+}  // namespace
+}  // namespace example
+}  // namespace totw
+```
+
+The vast majority of C++ users think that the using-directive is injecting names
+into the scope where it's declared. In the example above, that would be the
+scope of the function. In reality, the names are injected into the nearest
+common ancestor of the target namespace (`::testing`) and the usage namespace
+(`::totw::example::anomymous`). In our example, **that's the global namespace**!
+
+Thus, the code is roughly equivalent to the following:
+
+```c++
+using ::testing::Expectation;
+using ::testing::Sequence;
+using ::testing::UnorderedElementsAre;
+...
+// many, many more symbols are injected into the global namespace
+
+namespace totw {
+namespace example {
+namespace {
+
+TEST(MyTest, UsesUsingDirectives) {
+  Sequence seq; // ::testing::Sequence
+  WallTimer timer; // ::WallTimer
+  ...
+}
+
+} // namespace
+} // namespace example
+} // namespace totw
+```
+
+Let's see what kind of changes can break this code:
+
+*   If anyone defines `::totw::Sequence` or `::totw::example::Sequence`, `seq`
+    will now refer to that entity instead of `::testing::Sequence`.
+*   If anyone defines `::Sequence`, the definition of `seq` will fail to
+    compile, as the reference to the name `Sequence` will be ambiguous.
+    `Sequence` could mean `::testing::Sequence` or `::Sequence`, and the
+    compiler doesn't know which one you wanted.
+*   If anyone defines `::testing::WallTimer`, the definition of `timer` will
+    fail to compile.
+
+Thus, a single using-directive in a function scope has placed naming
+restrictions on symbols in `::testing`, `::totw`, `::totw::example`, and the
+global namespace. **Allowing this using-directive, even if only in function
+scope, creates ample opportunities for name clashes in the global and other
+namespaces.**
+
+If that example doesn't look fragile enough, consider this:
+
+```c++
+namespace totw {
+namespace example {
+namespace {
+
+TEST(MyTest, UsesUsingDirectives) {
+  using namespace ::testing;
+  EXPECT_THAT(..., proto::Partially(...)); // ::testing::proto::Partially
+  ...
+}
+
+} // namespace
+} // namespace example
+} // namespace totw
+```
+
+This using-directive has basically **introduced a namespace alias** `proto` in
+the global namespace, roughly equivalent to the following:
+
+<pre class="prettyprint lang-cpp bad-code">
+namespace proto = ::testing::proto;
+
+namespace totw {
+namespace example {
+namespace {
+
+TEST(MyTest, UsesUsingDirectives) {
+  EXPECT_THAT(..., proto::Partially(...)); // ::testing::proto::Partially
+  ...
+}
+
+} // namespace
+} // namespace example
+} // namespace totw
+</pre>
+
+The test will keep compiling until a header defining namespace `::proto`,
+`::totw::proto`, or `::totw::example::proto` gets included transitively. At that
+point in time, `proto::Partially` becomes ambiguous, and the test stops
+compiling. This ties into the style guide's rules on namespace naming: avoid
```

Comment:
If we're worried about brevity, the sentence starting "This ties into" can probably be deleted without losing any information.